contracts/contracts/trees/LeanIMT.sol

```diff
@@ -0,0 +1,348 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import { PoseidonT3 } from "../crypto/PoseidonT3.sol";
+
+struct LeanIMTData {
+  // Tracks the current number of leaves in the tree.
+  uint256 size;
+  // Represents the current depth of the tree, which can increase as new leaves are inserted.
+  uint256 depth;
+  // A mapping from each level of the tree to the node value of the last even position at that level.
+  // Used for efficient inserts, updates and root calculations.
+  mapping(uint256 => uint256) sideNodes;
+  // A mapping from leaf values to their respective indices in the tree.
+  // This facilitates checks for leaf existence and retrieval of leaf positions.
+  mapping(uint256 => uint256) leaves;
+}
+
+error WrongSiblingNodes();
+error LeafGreaterThanSnarkScalarField();
+error LeafCannotBeZero();
+error LeafAlreadyExists();
+error LeafDoesNotExist();
+
+/// @title Lean Incremental binary Merkle tree.
+/// @dev The LeanIMT is an optimized version of the BinaryIMT.
+/// This implementation eliminates the use of zeroes, and make the tree depth dynamic.
+/// When a node doesn't have the right child, instead of using a zero hash as in the BinaryIMT,
+/// the node's value becomes that of its left child. Furthermore, rather than utilizing a static tree depth,
+/// it is updated based on the number of leaves in the tree. This approach
+/// results in the calculation of significantly fewer hashes, making the tree more efficient.
+library InternalLeanIMT {
+  /// @notice The scalar field
+  uint256 internal constant SNARK_SCALAR_FIELD =
+    21888242871839275222246405745257275088548364400416034343698204186575808495617;
+
+  /// @dev Inserts a new leaf into the incremental merkle tree.
+  /// The function ensures that the leaf is valid according to the
+  /// constraints of the tree and then updates the tree's structure accordingly.
+  /// @param self: A storage reference to the 'LeanIMTData' struct.
+  /// @param leaf: The value of the new leaf to be inserted into the tree.
+  /// @return The new hash of the node after the leaf has been inserted.
+  function _insert(LeanIMTData storage self, uint256 leaf) internal returns (uint256) {
+    if (leaf >= SNARK_SCALAR_FIELD) {
+      revert LeafGreaterThanSnarkScalarField();
+    } else if (leaf == 0) {
+      revert LeafCannotBeZero();
+    } else if (_has(self, leaf)) {
+      revert LeafAlreadyExists();
+    }
+
+    uint256 index = self.size;
+
+    // Cache tree depth to optimize gas
+    uint256 treeDepth = self.depth;
+
+    // A new insertion can increase a tree's depth by at most 1,
+    // and only if the number of leaves supported by the current
+    // depth is less than the number of leaves to be supported after insertion.
+    if (2 ** treeDepth < index + 1) {
```

Comment:
Let's use shift `<<` instead of `**`

Reply:
Do you want to import LeanIMT from npm or just copy the file and change it?

Reply:
Some interesting things related to this update:

- Install `@zk-kit/lean-imt.sol` package in maci/contracts
- Update Maci.sol -> `import { LeanIMT, LeanIMTData } from "@zk-kit/lean-imt.sol/LeanIMT.sol";`
- It's required to install `poseidon-solidity` for PoseidonT3 that is used in the LeanIMT.sol
- Conflict is occurred between npm poseidon-solidity/PoseidonT3 and local crypto/PoseidonT3
- Renamed local crypto/PoseidonT3 to `crypto/PoseidonT3Lib`
- During deploy (`deployMaci`) is required to put the contract's address (the same approach as for poseidonT3/T4/T5/T6)
```
export interface MACILibraryAddresses {
  ["@zk-kit/lean-imt.sol/LeanIMT.sol:LeanIMT"]: string;
  ["contracts/crypto/PoseidonT3Lib.sol:PoseidonT3Lib"]: string;
  ["contracts/crypto/PoseidonT5.sol:PoseidonT5"]: string;
  ["contracts/crypto/PoseidonT6.sol:PoseidonT6"]: string;
  ["contracts/crypto/PoseidonT4.sol:PoseidonT4"]: string;
}
```

Reply:
mm is this a blocker or does it work? if that's a blocker let's go with copy/pasted version and we'll sort it out internally later

Reply:
Unfortunately, it is blocked without the contract's address. I'll leave that for later.

Reply:
ok no worries, let's copy paste then thank you